proxy_agent/src/key_keeper.rs

```diff
@@ -147,72 +138,69 @@ fn poll_secure_channel_status(
                         None => err_string,
                     }
                 );
-                unsafe {
-                    *STATUS_MESSAGE = message.to_string();
-                }
+                key_keeper_wrapper::set_status_message(shared_state.clone(), message.to_string());
                 logger::write_warning(message);
                 continue;
             }
         };
-        let mut guid;
-        match &status.keyGuid {
-            Some(id) => guid = id.to_string(),
-            None => guid = String::new(),
-        }
-
         logger::write_information(format!("Got key status successfully: {}.", status));
 
         let wireserver_rule_id = status.get_wireserver_rule_id();
-        let imds_rule_id = status.get_imds_rule_id();
-        unsafe {
-            if wireserver_rule_id != *WIRESERVER_RULE_ID {
-                logger::write_warning(format!(
-                    "Wireserver rule id changed from {} to {}.",
-                    *WIRESERVER_RULE_ID, wireserver_rule_id
-                ));
-                *WIRESERVER_RULE_ID = wireserver_rule_id.to_string();
-                proxy_authentication::set_wireserver_rules(status.get_wireserver_rules());
-            }
+        let imds_rule_id: String = status.get_imds_rule_id();
+        let current_wire_server_rule_id =
+            key_keeper_wrapper::get_wireserver_rule_id(shared_state.clone());
+        if wireserver_rule_id != current_wire_server_rule_id {
+            logger::write_warning(format!(
+                "Wireserver rule id changed from {} to {}.",
+                current_wire_server_rule_id, wireserver_rule_id
+            ));
+            key_keeper_wrapper::set_wireserver_rule_id(
+                shared_state.clone(),
+                wireserver_rule_id.to_string(),
+            );
+            proxy_authentication::set_wireserver_rules(status.get_wireserver_rules());
         }
-        unsafe {
-            if imds_rule_id != *IMDS_RULE_ID {
-                logger::write_warning(format!(
-                    "IMDS rule id changed from {} to {}.",
-                    *IMDS_RULE_ID, imds_rule_id
-                ));
-                *IMDS_RULE_ID = imds_rule_id.to_string();
-                proxy_authentication::set_imds_rules(status.get_imds_rules());
-            }
+
+        let current_imds_rule_id = key_keeper_wrapper::get_imds_rule_id(shared_state.clone());
```

Comment:
Similar to above, this introduces a race condition since the lock isn't held between reading the value and conditionally setting it.